### PR TITLE
Interview Scheduler - Single Page App

### PR DESCRIPTION
**Pull request by neverwannafly**
### Implemented
* Hash-based Router using Regular Expressions
* Token-based user authentication
* Interview creation and view
* User Resume upload
* Automated mail on interview creation

### TODO
* Interview edit and delete
* Display User Interviews
* Mails on interview edit
* Reorganize project, structure URL's and clean up code

### NOTE
* The front-end needs to be served from a different route. Ideally one can use vs-code live server. The configurations are already provided into the repo. 
* This previous pull request addresses the basic structure of the application. https://github.com/neverwannafly/RubyPlayground/pull/14. This pull request builds onto what I did in #14 

#### Reminder Emails
Screenshot 2020-05-20 at 6 35 55 PM
Screenshot 2020-05-20 at 6 36 02 PM

#### HOME SCREEN

### CODE REVIEW

File: interview-scheduler/src/services/SetUser.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  const email = user.email;
+  const token = user.token;
+  const username = user.username;
+
+  localStorage.setItem('userId', userId);
+  localStorage.setItem('name', name);
+  localStorage.setItem('email', email);
+  localStorage.setItem('token', token);
+  localStorage.setItem('username', username);
+}
```
Comment (ngudbhav) [inline interview-scheduler/src/services/SetUser.js:13]: Setting this level of sensitive information in LocalStorage can cause security issues. Can/Should we directly bind user object to window global object?

File: interview-scheduler/src/services/Config.js [hunk 2]
```diff
@@ -0,0 +1 @@
+export const ServerPreifx = 'http://localhost:3000';
```
Comment (ngudbhav) [inline interview-scheduler/src/services/Config.js:1]: Very good idea to use a single source of truth. I would say a better implementation of the same can be directly from the backend itself. I believe rails expose a `env_info` object for this. Can you please check this?

File: interview-scheduler/src/services/GetUser.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  return {
+    userId: localStorage.getItem('userId'),
+    name: localStorage.getItem('name'),
+    email: localStorage.getItem('email'),
+    token: localStorage.getItem('token'),
+    username: localStorage.getItem('username'),
+  }
+};
+
+export default GetUser;
```
Comment (ngudbhav) [inline interview-scheduler/src/services/GetUser.js:11]: We may all together remove this file if the user object is in the window object itself.

File: interview-scheduler/src/views/components/ViewInterview.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+              <a id="_delete_int_id" class="btn btn-outline-danger" data-confirm="Are you sure?" rel="nofollow" data-method="delete" href="#">Delete</a>
+              <button type="button" class="btn btn-outline-secondary" data-dismiss="modal">Close</button>
+            </div>
+          </div>
+        </div>
+      </div>
+    `;
+    return view;
+  },
+  postRender: async() => {
```
Comment (ngudbhav) [inline interview-scheduler/src/views/components/ViewInterview.js:66]: What is the need to make this an async function?

File: interview-scheduler/src/views/pages/Index.js [hunk 5]
```diff
@@ -1,58 +1,77 @@
...
+    const userData = GetUser();
 
     const calendar = new FullCalendar.Calendar(calendarEl, {
       plugins: ['timeGrid'],
       events: {
-        url: 'http://localhost:3000/api/interviews/fetch',
+        url: `${ServerPreifx}/api/interviews/fetch`,
+        extraParams: {
+          user_id: userData.userId,
+          token: userData.token,
```
Comment (ngudbhav) [inline interview-scheduler/src/views/pages/Index.js:35]: Why do we need this parameters sent from client. We have this data in backend right?

File: interview-scheduler/src/views/components/CreateInterview.js [hunk 6]
```diff
@@ -52,65 +55,36 @@ let Modals = {
...
+    const createInterview = document.getElementById('create_interview');
+    createInterview.addEventListener('submit', event => {
+      event.preventDefault();
+      const url = `${ServerPreifx}/interviews`;
+      const userData = GetUser();
+      $.ajax({
+        url: url,
+        data: {
+          user_id: userData.userId,
+          token: userData.token,
```
Comment (ngudbhav) [inline interview-scheduler/src/views/components/CreateInterview.js:71]: ^^ same here. I believe we have this data in backend

File: interview-scheduler/app/controllers/application_controller.rb [hunk 7]
```diff
@@ -13,18 +15,20 @@ def generate_token
...
   def current_user 
-    if session[:user_id] 
-      @_current_user ||= User.find_by(id: session[:user_id]) 
-    else
-      @_current_user = nil
+    if params[:user_id] && params[:token]
+      @_current_user ||= User.find_by(:id => params[:user_id], :token => params[:token])
     end
     return @_current_user
   end
```
Comment (ngudbhav) [inline interview-scheduler/app/controllers/application_controller.rb:23]: This method will serve the purpose of finding the logged in user. Therefore, we do not need the parameters from the frontend.

**Comment by ngudbhav** [comment]
Excellent work with the implementation. Few nits here and there. Nothing compulsory. Discuss if any doubt

File: interview-scheduler/src/services/Config.js (hunk 2 above)
Comment (neverwannafly) [reply interview-scheduler/src/services/Config.js:1]: I will read about it and try to incorporate in my next PR

File: interview-scheduler/src/services/SetUser.js (hunk 1 above)
Comment (neverwannafly) [reply interview-scheduler/src/services/SetUser.js:13]: I'll use window object

**Pull request merged**